docs/architecture/adr-027-deterministic-protobuf-serialization.md

```diff
@@ -45,11 +46,18 @@ with the following additions:
    must be omitted
 4. `repeated` fields of scalar numeric types must use
    [packed encoding](https://developers.google.com/protocol-buffers/docs/encoding#packed)
-   by default.
-5. Variant encoding of integers must not be longer than needed.
+5. Variant encoding must not be longer than needed:
+    1. The highest 6 bits for 10-byte varints must be `0`
+    1. No trailing zero bytes (in little endian, i.e. no leading zeroes in big
+       endian). Per rule 3 above, the default value of `0` must be omitted, so
+       this rule does not apply in such cases.
+    1. Highest 4 bytes of 8 must be `0` for 32-bit ints
```

Comment:
Are you referring to `fixed32` or 32-bit varints? In both cases, I actually also don't understand this rule 😅

Reply:
Implementation: https://github.com/lazyledger/protobuf3-solidity-lib/blob/v0.4.2/contracts/ProtobufLib.sol#L196-L200

Varints. As an example, `FF FF FF FF FF FF FF FF FF 01` is the maximum `uint64` value, and a valid varint. When casted to a `uint32`, the highest 4 bytes will be truncated, even if they're non-zero. Therefore non-zero bytes should be forbidden, since it's extra space that does nothing functionally.

Reply:
> When casted to a uint32

Why is casting to uint32 relevant to this ADR?

Reply:
Clarified in 6a6dbbb.